src/main.rs

```diff
@@ -14,137 +12,265 @@ fn log_2(x: u32) -> u32 {
+    shift_and_apply2(img, 0, distance, compare);
 }
 
-fn stretch<I, P, S>(img: I, border: u32) -> ImageBuffer<P, Vec<S>>
+fn subtract(img1: &mut GrayImage, img2: &GrayImage) {
+    apply2(img1, img2, |a, b| a.saturating_sub(b))
+}
+
+fn center_and_stretch<P, S>(img: &mut ImageBuffer<P, Vec<S>>, border: u32)
 where
-    I: GenericImageView<Pixel = P>,
     P: Pixel<Subpixel = S> + 'static,
     S: Primitive + 'static,
 {
-    let w = img.width();
-    let h = img.height();
-    let mut result = ImageBuffer::new(w + 2 * border - 1, h + 2 * border - 1);
-    replace(&mut result, &img, border, border);
-    let rw = result.width();
-    let rh = result.height();
+    let (width, height) = img.dimensions();
+    let (area_width, area_height) = (width - 2 * border, height - 2 * border);
+    let area = Rect {
+        x: 0,
+        y: 0,
+        width: area_width,
+        height: area_height,
+    };
+    img.copy_within(area, border, border);
     let top = Rect {
         x: 0,
         y: border,
-        width: rw,
+        width,
         height: 1,
     };
     let bottom = Rect {
         x: 0,
-        y: border + h - 1,
-        width: rw,
+        y: border + area_height - 1,
+        width,
         height: 1,
     };
     let left = Rect {
         x: border,
         y: 0,
         width: 1,
-        height: rh,
+        height,
     };
     let right = Rect {
-        x: border + w - 1,
+        x: border + area_width - 1,
         y: 0,
         width: 1,
-        height: rh,
+        height,
     };
     for y in 0..border {
-        result.copy_within(top, 0, y);
-        result.copy_within(bottom, 0, rh - y - 1);
+        img.copy_within(top, 0, y);
+        img.copy_within(bottom, 0, height - y - 1);
     }
     for x in 0..border {
-        result.copy_within(left, x, 0);
-        result.copy_within(right, rw - x - 1, 0);
+        img.copy_within(left, x, 0);
+        img.copy_within(right, width - x - 1, 0);
+    }
+}
+
+fn equalize(value: u8, black: u8, range: u8) -> u8 {
+    255.min(255 * value.saturating_sub(black) as u32 / range as u32) as u8
+}
+
+fn local_equalize(
+    img: &mut GrayImage,
+    local_black: GrayImage,
+    local_range: GrayImage,
+    debug_mode: bool,
+) {
+    save_debug_image(&img, "corrected.unequalized.png".to_string(), debug_mode);
+    apply3(
+        img,
+        &local_black,
+        &local_range,
+        |img_pixel, black, range| equalize(img_pixel, black, range),
+    )
+}
+
+fn save_debug_image(img: &GrayImage, name: String, debug_mode: bool) -> () {
+    if debug_mode {
+        img.save(format!("/tmp/{}", name)).unwrap();
+    }
+}
+
+fn brighter(a: u8, b: u8) -> u8 {
+    a.max(b)
+}
+
+fn darker(a: u8, b: u8) -> u8 {
+    a.min(b)
+}
+
+fn get_distribution(img: &GrayImage) -> [u32; 256] {
+    let mut result = [0; 256];
+    for p in img.pixels() {
+        result[p[0] as usize] += 1;
+    }
+    for i in 0..256 {
+        result[i] = ((result[i] as f32) + 1f32).log(1.1) as u32
```

Comment:
Why `.log(1.1)`?

Reply:
I experimented a bit, and `.log(2)` would give so small integers that smoothing lost too much detail and the local extrema were lost inside flat tops.

Not converting to logarithmic scale also in my opinion bent the local extrema detection too much towards the center. The value of 1.1 seemed to work well, although we could use a much lower log base and still be guaranteed to get results which fits in a `u32`. This of course depends on image size.

But yes, it would be good to understand why log base (or non log) affects the result of this algorithm as it does. Or what would be a better but equally well performing algorithm. Of course this one is just something an amateur non-mathematician came up with...

Reply:
Yeah it would be nice to plot these histograms with and without log. Hopefully when we have the wasm version it will be easy to see the difference